Fix panic() on aerofc

lucasdemarchi opened the pull request:
This should fix #8416

lucasdemarchi commented:
Just tested on a real vehicle and it works if the second commit stays... i.e. there are probably more parameters it's not finding. Since the parameter is actually valid, I think only the second commit is right (I see no reason for keeping the panic() there), but we need to find out why it doesn't find the parameter.

lucasdemarchi commented:
> >but we need to find out why it doesn't find the parameter.
> 
> This looks like something is wrong in AeroFC and we should find the problem instead of simply by-passing it.

I don't think it's exclusive to aerofc, but as I said I agree we need to find why it's not loading valid parameters.

lucasdemarchi commented:
> I don't think it's exclusive to aerofc, but as I said I agree we need to find why it's not loading valid parameters.

tldr: it seems defaults are broken for dynamically loaded params.

Long version:
yep, it's not and can be reproduced in sitl:

```
./waf configure --board sitl
./waf cotper
 gdb ./build/sitl/bin/arducopter --model x --defaults Tools/Frame_params/intel-aero-rtf.param
....
invalid param ATC_ACCEL_P_MAX in defaults file
invalid param ATC_ACCEL_R_MAX in defaults file
invalid param ATC_ACCEL_Y_MAX in defaults file
invalid param ATC_ANG_PIT_P in defaults file
invalid param ATC_ANG_RLL_P in defaults file
...
Loaded defaults from Tools/Frame_params/intel-aero-rtf.param
AP_Param: Invalid param BRD_SAFETY_MASK in defaults file
^C
(gdb) bt
#0  0x00000000004bd2e0 in AP_HAL::panic(char const*, ...) ()
#1  0x000000000045ad39 in AP_Param::count_defaults_in_file(char const*, unsigned short&, bool) ()
#2  0x000000000045aef5 in AP_Param::load_defaults_file(char const*, bool) [clone .part.11] ()
#3  0x000000000045b430 in AP_Param::reload_defaults_file(bool) ()
#4  0x0000000000426f4f in Copter::allocate_motors() ()
#5  0x000000000042791e in Copter::init_ardupilot() ()
#6  0x000000000040656c in Copter::setup() ()
#7  0x00000000004b3ffd in HAL_SITL::run(int, char* const*, AP_HAL::HAL::Callbacks*) const ()
#8  0x0000000000402eee in main ()
```

AFAICS it's broken for the allocated params since it's not updating `_num_vars`. The first prints are when `load_all(false)` is called. The last ones cause the call to reload defaults will pass true as argument.  I guess the same could be reproduced on on boards using apj_tool, but for those @tridge did exactly what I'm doing here in commit 51512ed7c05ce9c7911d13acfacadcdfa50bbde3.

tridge commented:
SITL doesn't have a BRD_SAFETY_MASK parameter, so it failing on that parameter is expected.
I don't mind changing the error to a warning

tridge commented on libraries/AP_Param/AP_Param.h:
```diff
@@ -327,7 +327,7 @@ class AP_Param
     /// reoad the hal.util defaults file. Called after pointer parameters have been allocated
     ///
-    static void reload_defaults_file(bool panic_on_error=true);
+    static void reload_defaults_file(bool panic_on_error=false);
```
I think we can change it to a warning on hal.console

lucasdemarchi commented:
True, I missed that. It doesn't explain why it's failing on aerofc for a valid param with default loaded from /etc though. It seems I will need to debug this more.

tridge commented:
> True, I missed that. It doesn't explain why it's failing on aerofc for a valid param with default loaded from /etc 

it will be related to the two-pass nature of some parameters. Parameters that are part of a dynamically loaded class that only exist once you set a parameter are tricky. There is deliberately two calls into AP_Param to cope with that
I suspect the parameter you hit doesn't exist until after FRAME_CLASS is set

lucasdemarchi commented:
@tridge I thought so, but copter only checks the value of frame class after it... I went ahead and got all the params in which it's failing.

```diff
diff --git a/Tools/Frame_params/intel-aero-rtf.param b/Tools/Frame_params/intel-aero-rtf.param
index 9402f9436d..8d92b344c1 100644
--- a/Tools/Frame_params/intel-aero-rtf.param
+++ b/Tools/Frame_params/intel-aero-rtf.param
@@ -35,14 +35,11 @@ FLTMODE1,5
 FLTMODE4,2
 FRAME_CLASS,1
 FRAME_TYPE,1
-FS_BATT_ENABLE,2
-FS_BATT_VOLTAGE,10.5
 FS_THR_ENABLE,1
 LOG_BACKEND_TYPE,2
 MOT_BAT_VOLT_MAX,12.6
 MOT_BAT_VOLT_MIN,10.5
 MOT_THST_HOVER,0.35
-RC_FEEL_RP,100
 RC1_REVERSED,1
 RC1_DZ,30
 RC2_REVERSED,1
@@ -54,8 +51,4 @@ SERVO2_FUNCTION,34
 SERVO3_FUNCTION,35
 SERVO4_FUNCTION,36
 WPNAV_ACCEL,250
-WPNAV_ACCEL_Z,100
-WPNAV_LOIT_JERK,4000
-WPNAV_LOIT_MAXA,650
-WPNAV_LOIT_MINA,150
-WPNAV_LOIT_SPEED,700
\ No newline at end of file
+WPNAV_ACCEL_Z,100
\ No newline at end of file
```

So, In my first try I may have been mislead by the first pass and it worked simply because I removed the panic().  Now checking this parameters, it seems like at least some were removed/renamed.

```sh
$ git grep -l FS_BATT_ENABLE -- Tools/Frame_params/ | wc -l 
12
```

I think we should be really careful when renaming params. Parameter files from Tools/Frame_params/ on **master** branch should apply to the firmware in the **master** branch.  I remember someone saying this is not the case because Mission Planner uses that files in the last stable branch... IMO it should be using the files from the stable branch then, not from master. @meee1 can you confirm this is the case?

Then anytime a param is renamed or removed a simple "git grep" can show all the places in which it needs to be changed.

lucasdemarchi commented:
Just pushed a commit with changes to parameters. I removed the WPNAV params since I'm not sure what to do.... apparently they were not only renamed but changed meaning, too. RC_FEEL_RP changed the scale and I added the new value.

OXINARF commented:
I think we need to make the warnings more prominent. An ignored parameter can be very important! For example, the battery ones that were changed recently, including failsafe, can bring a vehicle down. Upgrading from one version to another will work if the parameters are saved in flash (conversion code will run), but not with a defaults file.

tridge commented:
more prominent is fine, but only if it is correct. If the warning is given on the first pass then it may not be correct, as dynamic parameters trees are not initialised yet

OXINARF commented:
Yes, that is needed.

lucasdemarchi commented:
> I think we should only printf() on the 2nd pass,

agreed, will do.

> I think we need to make the warnings more prominent. An ignored parameter can be very important! For example, the battery ones that were changed recently, including failsafe, can bring a vehicle down. Upgrading from one version to another will work if the parameters are saved in flash (conversion code will run), but not with a defaults file.

I think what we really need is people to do a `git grep` in the codebase and change the places referring that param in the same PR in which they are being renamed.  The current code for converting old names to new ones works only for transferring the value saved in eprom/file to the new place. However the handling for default values doesn't.

lucasdemarchi commented:
> I think we need to make the warnings more prominent

btw, I thought about usin hal.console() here, would that help?

OXINARF commented:
> I think what we really need is people to do a git grep in the codebase and change the places referring that param in the same PR in which they are being renamed. The current code for converting old names to new ones works only for transferring the value saved in eprom/file to the new place. However the handling for default values doesn't.

I can start asking people to do that, but it's not impossible that it's missed. Also, users might not be using a parameter file from the repo, but their own. It's also possible that an user updates the firmware version and doesn't update the parameter file.
So although I agree that updating the parameter files in repo is worth it, it's not going to fix the problem.

> btw, I thought about usin hal.console() here, would that help?

I'm not entirely sure that's enough. How many users are going to catch it there? Particularly in Linux boards (where I think it's more likely that defaults file is used), is the console usually connected to a remote link?

lucasdemarchi commented:
@OXINARF I decided to send both to console and debug terminal. It's convenient so we more people can see the warning we want to show. It also kind of matches what we were doing before with the panic().

@tridge @OXINARF please take a look again

lucasdemarchi commented:
I'm applying this as it unbreaks aerofc on master branch and from my tests it's an improvements for all platforms (it doesn't panic() on the first parameter pass)... we can always revert or fixup if any bug is found by late reviews.
